packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+          scriptsArr[i].onPointerExit();
+        }
+      }
+      input.enteredEntity = curEntity;
+    }
+    /** Check whether down up and click events are triggered. */
+    while (evtTypeList > 1) {
+      switch (evtTypeList % 2) {
+        case PointerEventType.Down:
+          if (curEntity) {
+            for (let j = 0; j < curEntityScriptsLen; j++) {
+              curEntityScripts[j].onPointerDown();
+            }
+          }
+          prePressedEntity = curEntity;
+          break;
+        case PointerEventType.Up:
+          if (prePressedEntity) {
+            if (prePressedEntity == curEntity) {
+              for (let j = 0; j < curEntityScriptsLen; j++) {
+                const script = curEntityScripts[j];
+                script.onPointerUp();
+                script.onPointerClick();
+              }
+            } else {
+              const preScripts = prePressedEntity._scripts._elements;
+              for (let j = preScripts.length - 1; j >= 0; j--) {
+                preScripts[j].onPointerUp();
+              }
+            }
+          }
+          prePressedEntity = null;
+          break;
+      }
+      evtTypeList >>= 1;
+    }
+    input.pressedEntity = prePressedEntity;
+  }
+
+  /**
+   * Get the Entity to which the ray is cast.
+   * @param x - The X coordinate of the pointer on the screen, specified in normalized
+   * @param y - The Y coordinate of the pointer on the screen, specified in normalized
+   * @returns The Entity to which the ray is cast
+   */
+  private _pointerRaycast(x: number, y: number): Entity {
+    const { _engine: engine, _canvas: canvas } = this;
+    /** Convert screen coordinates to viewport coordinates. */
+    x = (x - canvas.offsetLeft) / canvas.clientWidth;
+    y = (y - canvas.offsetTop) / canvas.clientHeight;
+    const cameras = engine.sceneManager.activeScene._activeCameras;
+    for (let i = cameras.length - 1; i >= 0; i--) {
+      const camera = cameras[i];
+      if (!camera.enabled || camera.renderTarget) {
+        continue;
+      }
+      const { x: vpX, y: vpY, z: vpW, w: vpH } = camera.viewport;
+      if (x >= vpX && y >= vpY && x - vpX <= vpW && y - vpY <= vpH) {
+        const { _tempHitResult, _tempPoint } = InputManager;
+        _tempPoint.setValue((x - vpX) / vpW, (y - vpY) / vpH);
+        // TODO: Only check which colliders have listened to the input.
+        return engine.physicsManager.raycast(
+          camera.viewportPointToRay(_tempPoint, InputManager._tempRay),
+          _tempHitResult
+        )
+          ? _tempHitResult.collider.entity
+          : null;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Update pointers info.
+   * @param type - Pointer change type
+   * @param pointerId - Pointer ID
+   * @param x - The x coordinate of Pointer
+   * @param y - The y coordinate of Pointer
+   */
+  private _changePointer(type: PointerOperationType, pointerId: number, x?: number, y?: number): void {
```

Comment:
Move `_changePointer` to a class  like `PointerManager`  to handle CRUD.

Reply:
Done.